# Putting it all together
The goal of this notebook is to add Ballpark factors, team hitting data, team bullpen data and our target variable to the dataset and prepare it for EDA.

In [311]:
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests
import unidecode
from datetime import datetime
from datetime import timedelta

In [312]:
def chatGPT(text):
  url = "https://api.openai.com/v1/completions"
  headers = {
  "Content-Type": "application/json",
  "Authorization": "Bearer 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",
  }
  data = { 
  "model": "text-davinci-003",
  "prompt": text,
  "max_tokens": 4000,
  "temperature": 1.0,
  }
  response = requests.post(url, headers=headers, json=data)
  output = response.json()["choices"][0]["text"]
  
  return print(output)

# Main dataframe
We'll start by bringing in our main dataframe.

In [313]:
main_df = pd.read_csv('main_SP_features.csv')

In [314]:
main_df['Date'] = pd.to_datetime(main_df['Date'])

In [315]:
main_df.head()

Unnamed: 0       Date  game_id A_Team  A_Score  A_Win  A_Team_Wins  \
0           0 2021-04-06       59    HOU        4      1          4.0   
1           3 2021-04-06       67    LAD        5      1          4.0   
2           5 2021-04-07       73    TBR        2      0          2.0   
3           6 2021-04-07       74    MIL        4      1          2.0   
4           7 2021-04-07       75    PIT        4      0          1.0   

   A_Wins_L10  A_Wins_L30  A_Py_Wins  ...  H_Py_Wins H_Run_Diff  H_Avg_Outs  \
0         4.0         4.0        4.0  ...        3.0        3.0        18.0   
1         4.0         4.0        3.0  ...        0.0      -33.0        15.0   
2         2.0         2.0        2.0  ...        2.0       -3.0        15.0   
3         2.0         2.0        2.0  ...        3.0        1.0         9.0   
4         1.0         1.0        1.0  ...        4.0       24.0        10.0   

   H_BABIP H_FIP H_xFIP  H_WAR  H_WHIP H_SIERA  H_RS/9  
0    0.294  3.84   3.14    0.1    1.17    3.39     3.0  
1    0.250  3.73   5.39    0.1    1.13    5.55     0.0  
2    0.286  2.23   3.89    0.2    0.94    3.81     0.0  
3    0.375  7.84   4.68   -0.1    2.33    5.28     6.0  
4    0.412  9.77   7.99   -0.1    3.00    6.50     8.1  

[5 rows x 36 columns]

In [316]:
main_df.tail()

Unnamed: 0       Date  game_id A_Team  A_Score  A_Win  A_Team_Wins  \
4362        4448 2022-10-05     4854    LAA        2      0         73.0   
4363        4449 2022-10-05     4855    STL        3      0         93.0   
4364        4450 2022-10-05     4856    SFG        8      1         80.0   
4365        4451 2022-10-05     4857    DET        4      0         66.0   
4366        4452 2022-10-05     4858    NYY        2      0         99.0   

      A_Wins_L10  A_Wins_L30  A_Py_Wins  ...  H_Py_Wins H_Run_Diff  \
4362         7.0        16.0       75.0  ...       58.0     -203.0   
4363         5.0        17.0       95.0  ...       57.0     -228.0   
4364         6.0        17.0       82.0  ...       86.0       52.0   
4365         7.0        16.0       63.0  ...       88.0       66.0   
4366         6.0        20.0      106.0  ...       77.0      -38.0   

      H_Avg_Outs  H_BABIP H_FIP H_xFIP  H_WAR  H_WHIP H_SIERA  H_RS/9  
4362       13.67    0.320  4.67   4.24    0.1    1.37    3.88    5.53  
4363       13.43    0.279  3.82   4.20    0.6    1.29    4.00    5.37  
4364        6.00    0.282  4.89   3.51   -0.3    1.22    3.25    5.97  
4365       16.81    0.274  5.15   4.89   -0.1    1.33    5.00    5.01  
4366       14.85    0.257  5.23   4.87    0.0    1.35    4.86    4.79  

[5 rows x 36 columns]

In [317]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4367 entries, 0 to 4366
Data columns (total 36 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Unnamed: 0   4367 non-null   int64         
 1   Date         4367 non-null   datetime64[ns]
 2   game_id      4367 non-null   int64         
 3   A_Team       4367 non-null   object        
 4   A_Score      4367 non-null   int64         
 5   A_Win        4367 non-null   int64         
 6   A_Team_Wins  4367 non-null   float64       
 7   A_Wins_L10   4367 non-null   float64       
 8   A_Wins_L30   4367 non-null   float64       
 9   A_Py_Wins    4367 non-null   float64       
 10  A_Run_Diff   4367 non-null   float64       
 11  H_Team       4367 non-null   object        
 12  A_Avg_Outs   4367 non-null   float64       
 13  A_BABIP      4367 non-null   float64       
 14  A_FIP        4367 non-null   object        
 15  A_xFIP       4367 non-null   object        
 16  A_WAR 

In [318]:
main_df = main_df.drop(columns = ['Unnamed: 0'])

# A little glitch
It looks like we duplicated the home team column 'H_Team' in the main dataframe. Let's just make sure it really is a duplication before dropping one of them.<br>

And we see that the columns are identical in all 4367 rows of the main dataframe.

In [319]:
len(main_df[main_df['H_Team'] == main_df['H_Team.1']])

4367

In [320]:
len(main_df[main_df['H_Team'] != main_df['H_Team.1']])

0

In [321]:
main_df = main_df.drop(columns = ['H_Team'])

In [322]:
main_df = main_df.rename(columns = {'H_Team.1': 'H_Team'})

In [323]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4367 entries, 0 to 4366
Data columns (total 34 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4367 non-null   datetime64[ns]
 1   game_id      4367 non-null   int64         
 2   A_Team       4367 non-null   object        
 3   A_Score      4367 non-null   int64         
 4   A_Win        4367 non-null   int64         
 5   A_Team_Wins  4367 non-null   float64       
 6   A_Wins_L10   4367 non-null   float64       
 7   A_Wins_L30   4367 non-null   float64       
 8   A_Py_Wins    4367 non-null   float64       
 9   A_Run_Diff   4367 non-null   float64       
 10  A_Avg_Outs   4367 non-null   float64       
 11  A_BABIP      4367 non-null   float64       
 12  A_FIP        4367 non-null   object        
 13  A_xFIP       4367 non-null   object        
 14  A_WAR        4367 non-null   float64       
 15  A_WHIP       4367 non-null   float64       
 16  A_SIER

# BABIP
We have BABIP for starting pitchers and we'll also have team hitting BABIP when we bring in the hitting data. So we'll need to add an '_SP' suffix to BABIP before we bring in the hitting data.

In [324]:
main_df = main_df.rename(columns = {'A_BABIP': 'A_BABIP_SP', 'H_BABIP': 'H_BABIP_SP'})

# Dropping Toronto rows before July 30, 2021
We're going to merge in team hitting and team bullpen data before ballpark factors, but just to save a little computational time we're first going to shed data from all Blue Jays home games before July 30, 2021. The Blue Jays didn't play any games in their home stadium in 2021 until July 30 because of COVID, so the ballpark factors wouldn't be accurate for TOR home games before that date.

In [325]:
main_df = main_df[~((main_df['H_Team'] == 'TOR') & (main_df['Date'] < '2021-07-30'))]

In [326]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4366
Data columns (total 34 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   A_Team       4327 non-null   object        
 3   A_Score      4327 non-null   int64         
 4   A_Win        4327 non-null   int64         
 5   A_Team_Wins  4327 non-null   float64       
 6   A_Wins_L10   4327 non-null   float64       
 7   A_Wins_L30   4327 non-null   float64       
 8   A_Py_Wins    4327 non-null   float64       
 9   A_Run_Diff   4327 non-null   float64       
 10  A_Avg_Outs   4327 non-null   float64       
 11  A_BABIP_SP   4327 non-null   float64       
 12  A_FIP        4327 non-null   object        
 13  A_xFIP       4327 non-null   object        
 14  A_WAR        4327 non-null   float64       
 15  A_WHIP       4327 non-null   float64       
 16  A_SIER

# Breaking down into home and away rows
In order to merge in team hitting and bullpen data, we need to once again create one row for every team in every game.

In [327]:
away_cols = list(main_df.columns)[:18]

In [328]:
away_cols

['Date',
 'game_id',
 'A_Team',
 'A_Score',
 'A_Win',
 'A_Team_Wins',
 'A_Wins_L10',
 'A_Wins_L30',
 'A_Py_Wins',
 'A_Run_Diff',
 'A_Avg_Outs',
 'A_BABIP_SP',
 'A_FIP',
 'A_xFIP',
 'A_WAR',
 'A_WHIP',
 'A_SIERA',
 'A_RS/9']

In [329]:
chatGPT("How do I index different parts of a list in Python? For example, I want index items 0 and 1 and also 18 until the end.")

 

To index items 0 and 1 and also 18 until the end of the list in Python, you can use list slicing. List slicing allows you to specify a start and end index, as well as a step size (which defaults to 1). The syntax for this is [start:end:step]. In your example, the code would look something like this: 

my_list[0:2:1] + my_list[18::] 

This will return a new list with all the items from indices 0 and 1, as well as all the items from index 18 to the end of the list.


In [330]:
home_cols = list(main_df.columns)[:2] + list(main_df.columns)[18:]

In [331]:
home_cols

['Date',
 'game_id',
 'H_Team',
 'H_Score',
 'H_Win',
 'H_Team_Wins',
 'H_Wins_L10',
 'H_Wins_L30',
 'H_Py_Wins',
 'H_Run_Diff',
 'H_Avg_Outs',
 'H_BABIP_SP',
 'H_FIP',
 'H_xFIP',
 'H_WAR',
 'H_WHIP',
 'H_SIERA',
 'H_RS/9']

In [332]:
main_df_away = main_df[away_cols]

In [333]:
main_df_away.head()

Date  game_id A_Team  A_Score  A_Win  A_Team_Wins  A_Wins_L10  \
0 2021-04-06       59    HOU        4      1          4.0         4.0   
1 2021-04-06       67    LAD        5      1          4.0         4.0   
2 2021-04-07       73    TBR        2      0          2.0         2.0   
3 2021-04-07       74    MIL        4      1          2.0         2.0   
4 2021-04-07       75    PIT        4      0          1.0         1.0   

   A_Wins_L30  A_Py_Wins  A_Run_Diff  A_Avg_Outs  A_BABIP_SP A_FIP A_xFIP  \
0         4.0        4.0        25.0        18.0       0.200  1.84    3.6   
1         4.0        3.0        12.0        17.0       0.400  2.99   3.62   
2         2.0        2.0       -12.0        17.0       0.235  2.11    4.6   
3         2.0        2.0        -5.0        12.0       0.462  2.17   3.05   
4         1.0        1.0       -18.0         9.0       0.111  4.84    6.6   

   A_WAR  A_WHIP A_SIERA  A_RS/9  
0    0.2    0.50    3.83    4.50  
1    0.1    1.94    3.77    6.35  
2    0.2    0.71    4.70    1.59  
3    0.2    2.00    4.22    2.25  
4    0.0    1.33    7.41    9.00

In [334]:
main_df_away.tail()

Date  game_id A_Team  A_Score  A_Win  A_Team_Wins  A_Wins_L10  \
4362 2022-10-05     4854    LAA        2      0         73.0         7.0   
4363 2022-10-05     4855    STL        3      0         93.0         5.0   
4364 2022-10-05     4856    SFG        8      1         80.0         6.0   
4365 2022-10-05     4857    DET        4      0         66.0         7.0   
4366 2022-10-05     4858    NYY        2      0         99.0         6.0   

      A_Wins_L30  A_Py_Wins  A_Run_Diff  A_Avg_Outs  A_BABIP_SP A_FIP A_xFIP  \
4362        16.0       75.0       -44.0       17.85       0.294  2.44   2.66   
4363        17.0       95.0       137.0       12.33       0.330  5.03   4.94   
4364        17.0       82.0        12.0        3.00       0.311  3.35   4.52   
4365        16.0       63.0      -155.0       13.81       0.281  4.77   4.82   
4366        20.0      106.0       242.0       15.15       0.251  4.27   4.47   

      A_WAR  A_WHIP A_SIERA  A_RS/9  
4362    5.4    1.03    2.74    4.42  
4363    0.0    1.69    4.84    9.40  
4364    0.8    1.30    4.11    4.30  
4365    0.1    1.31    4.89    3.59  
4366    0.7    1.12    4.36    3.31

In [335]:
main_df_away.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4366
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   A_Team       4327 non-null   object        
 3   A_Score      4327 non-null   int64         
 4   A_Win        4327 non-null   int64         
 5   A_Team_Wins  4327 non-null   float64       
 6   A_Wins_L10   4327 non-null   float64       
 7   A_Wins_L30   4327 non-null   float64       
 8   A_Py_Wins    4327 non-null   float64       
 9   A_Run_Diff   4327 non-null   float64       
 10  A_Avg_Outs   4327 non-null   float64       
 11  A_BABIP_SP   4327 non-null   float64       
 12  A_FIP        4327 non-null   object        
 13  A_xFIP       4327 non-null   object        
 14  A_WAR        4327 non-null   float64       
 15  A_WHIP       4327 non-null   float64       
 16  A_SIER

In [336]:
main_df_home = main_df[home_cols]

In [337]:
main_df_home.head()

Date  game_id H_Team  H_Score  H_Win  H_Team_Wins  H_Wins_L10  \
0 2021-04-06       59    LAA      2.0    0.0          4.0         4.0   
1 2021-04-06       67    OAK      1.0    0.0          0.0         0.0   
2 2021-04-07       73    BOS      9.0    1.0          2.0         2.0   
3 2021-04-07       74    CHC      2.0    0.0          3.0         3.0   
4 2021-04-07       75    CIN     11.0    1.0          4.0         4.0   

   H_Wins_L30  H_Py_Wins  H_Run_Diff  H_Avg_Outs  H_BABIP_SP  H_FIP  H_xFIP  \
0         4.0        3.0         3.0        18.0       0.294   3.84    3.14   
1         0.0        0.0       -33.0        15.0       0.250   3.73    5.39   
2         2.0        2.0        -3.0        15.0       0.286   2.23    3.89   
3         3.0        3.0         1.0         9.0       0.375   7.84    4.68   
4         4.0        4.0        24.0        10.0       0.412   9.77    7.99   

   H_WAR  H_WHIP  H_SIERA  H_RS/9  
0    0.1    1.17     3.39     3.0  
1    0.1    1.13     5.55     0.0  
2    0.2    0.94     3.81     0.0  
3   -0.1    2.33     5.28     6.0  
4   -0.1    3.00     6.50     8.1

In [338]:
main_df_home.tail()

Date  game_id H_Team  H_Score  H_Win  H_Team_Wins  H_Wins_L10  \
4362 2022-10-05     4854    OAK      3.0    1.0         59.0         4.0   
4363 2022-10-05     4855    PIT      5.0    1.0         61.0         6.0   
4364 2022-10-05     4856    SDP      1.0    0.0         89.0         6.0   
4365 2022-10-05     4857    SEA      5.0    1.0         89.0         6.0   
4366 2022-10-05     4858    TEX      4.0    1.0         67.0         2.0   

      H_Wins_L30  H_Py_Wins  H_Run_Diff  H_Avg_Outs  H_BABIP_SP  H_FIP  \
4362        10.0       58.0      -203.0       13.67       0.320   4.67   
4363        12.0       57.0      -228.0       13.43       0.279   3.82   
4364        17.0       86.0        52.0        6.00       0.282   4.89   
4365        16.0       88.0        66.0       16.81       0.274   5.15   
4366         9.0       77.0       -38.0       14.85       0.257   5.23   

      H_xFIP  H_WAR  H_WHIP  H_SIERA  H_RS/9  
4362    4.24    0.1    1.37     3.88    5.53  
4363    4.20    0.6    1.29     4.00    5.37  
4364    3.51   -0.3    1.22     3.25    5.97  
4365    4.89   -0.1    1.33     5.00    5.01  
4366    4.87    0.0    1.35     4.86    4.79

In [339]:
main_df_home.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4366
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   H_Team       4327 non-null   object        
 3   H_Score      4327 non-null   float64       
 4   H_Win        4327 non-null   float64       
 5   H_Team_Wins  4327 non-null   float64       
 6   H_Wins_L10   4327 non-null   float64       
 7   H_Wins_L30   4327 non-null   float64       
 8   H_Py_Wins    4327 non-null   float64       
 9   H_Run_Diff   4327 non-null   float64       
 10  H_Avg_Outs   4327 non-null   float64       
 11  H_BABIP_SP   4327 non-null   float64       
 12  H_FIP        4327 non-null   float64       
 13  H_xFIP       4327 non-null   float64       
 14  H_WAR        4327 non-null   float64       
 15  H_WHIP       4327 non-null   float64       
 16  H_SIER

In [340]:
main_df_away = main_df_away.rename(columns = {'A_Team': 'Team'})
main_df_home = main_df_home.rename(columns = {'H_Team': 'Team'})

## Bringing in team hitting data

In [341]:
hitting_df = pd.read_csv('Hitting_Data_2021-22.csv')

In [342]:
hitting_df.head()

Unnamed: 0 Team    OPS  BABIP   wOBA  WAR    ISO        Date
0           0  KCR  1.132  0.414  0.471  0.8  0.237  2021-04-02
1           1  LAD  1.077  0.484  0.461  0.7  0.139  2021-04-02
2           2  TEX  0.950  0.625  0.424  0.7  0.075  2021-04-02
3           3  SDP  1.004  0.381  0.417  0.5  0.281  2021-04-02
4           4  ARI  0.992  0.333  0.411  0.4  0.375  2021-04-02

In [343]:
hitting_df.tail()

Unnamed: 0 Team    OPS  BABIP   wOBA  WAR    ISO        Date
10798        1315  CHW  0.659  0.272  0.288  1.2  0.149  2022-10-06
10799        1316  TEX  0.653  0.269  0.286  1.4  0.163  2022-10-06
10800        1317  ARI  0.647  0.271  0.283  2.5  0.146  2022-10-06
10801        1318  OAK  0.632  0.268  0.280  1.1  0.141  2022-10-06
10802        1319  CIN  0.609  0.250  0.272 -1.4  0.129  2022-10-06

In [344]:
hitting_df = hitting_df.drop(columns = ['Unnamed: 0'])

In [345]:
hitting_df['Date'] = pd.to_datetime(hitting_df['Date'])

In [346]:
hitting_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10803 entries, 0 to 10802
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Team    10803 non-null  object        
 1   OPS     10803 non-null  float64       
 2   BABIP   10803 non-null  float64       
 3   wOBA    10803 non-null  float64       
 4   WAR     10803 non-null  float64       
 5   ISO     10803 non-null  float64       
 6   Date    10803 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(5), object(1)
memory usage: 590.9+ KB


In [347]:
hitting_df[hitting_df['Date'] == '2021_07-12']

Empty DataFrame
Columns: [Team, OPS, BABIP, wOBA, WAR, ISO, Date]
Index: []

In [348]:
main_df_away = pd.merge(main_df_away, hitting_df, how = 'left', on = ['Date', 'Team'])

In [349]:
main_df_away.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4326
Data columns (total 23 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   Team         4327 non-null   object        
 3   A_Score      4327 non-null   int64         
 4   A_Win        4327 non-null   int64         
 5   A_Team_Wins  4327 non-null   float64       
 6   A_Wins_L10   4327 non-null   float64       
 7   A_Wins_L30   4327 non-null   float64       
 8   A_Py_Wins    4327 non-null   float64       
 9   A_Run_Diff   4327 non-null   float64       
 10  A_Avg_Outs   4327 non-null   float64       
 11  A_BABIP_SP   4327 non-null   float64       
 12  A_FIP        4327 non-null   object        
 13  A_xFIP       4327 non-null   object        
 14  A_WAR        4327 non-null   float64       
 15  A_WHIP       4327 non-null   float64       
 16  A_SIER

In [350]:
chatGPT("How can I show only the rows of a dataframe that have missing values in pandas?")



You can use the Pandas function isnull() followed by the sum() function, set the axis to 1 (columns) and then filter the dataframe for any null values.

Example:

df[df.isnull().sum(axis=1) != 0]


In [351]:
rows_with_missing = main_df_away.isnull().any(axis = 1)

In [352]:
main_df_away[rows_with_missing]

Date  game_id Team  A_Score  A_Win  A_Team_Wins  A_Wins_L10  \
1102 2021-07-16     1348  SEA        6      1         48.0         6.0   
1103 2021-07-16     1349  CHC        5      1         44.0         2.0   
1104 2021-07-16     1350  TBR        7      1         53.0         6.0   
1105 2021-07-16     1351  HOU        7      1         55.0         7.0   
1106 2021-07-16     1352  MIL       11      1         53.0         4.0   
1107 2021-07-16     1353  LAD       10      1         56.0         6.0   
1108 2021-07-16     1354  BAL        2      0         28.0         3.0   
1109 2021-07-16     1355  BOS        4      1         55.0         5.0   
1110 2021-07-16     1356  CLE        4      0         45.0         3.0   
1111 2021-07-16     1359  NYM        1      0         48.0         6.0   
1112 2021-07-16     1360  SFG        7      1         57.0         7.0   
1113 2021-07-16     1362  SDP       24      1         53.0         4.0   
3352 2022-07-21     3815  SFG        6      0         48.0         7.0   
3353 2022-07-21     3816  TEX        8      1         41.0         4.0   

      A_Wins_L30  A_Py_Wins  A_Run_Diff  ...  A_xFIP  A_WAR A_WHIP A_SIERA  \
1102        18.0       40.0       -50.0  ...    4.35    1.8   1.21    4.54   
1103        11.0       42.0       -24.0  ...    4.34    0.7   1.25    4.47   
1104        16.0       54.0        85.0  ...    4.37    0.5   1.33    4.34   
1105        20.0       58.0       136.0  ...    4.09    1.5   1.23    4.33   
1106        18.0       51.0        46.0  ...    3.99    0.6   1.38    4.45   
1107        20.0       60.0       142.0  ...    3.46    2.3   1.07    3.49   
1108         7.0       32.0      -138.0  ...    4.61    0.2   1.68    4.44   
1109        18.0       51.0        57.0  ...    3.37    1.8   1.37    3.53   
1110        14.0       42.0       -16.0  ...    4.87   -0.2   1.50    4.41   
1111        15.0       45.0        10.0  ...    3.63    1.6   1.13    4.03   
1112        20.0       57.0       116.0  ...    2.97    1.2   1.23    3.39   
1113        16.0       55.0        75.0  ...    3.59    1.3   1.35    3.79   
3352        13.0       51.0        51.0  ...    3.10    3.8   1.11    3.15   
3353        12.0       45.0        -1.0  ...    3.24    1.7   1.13    3.38   

      A_RS/9  OPS BABIP  wOBA  WAR  ISO  
1102    5.26  NaN   NaN   NaN  NaN  NaN  
1103    6.34  NaN   NaN   NaN  NaN  NaN  
1104    4.55  NaN   NaN   NaN  NaN  NaN  
1105    4.86  NaN   NaN   NaN  NaN  NaN  
1106    3.95  NaN   NaN   NaN  NaN  NaN  
1107    7.62  NaN   NaN   NaN  NaN  NaN  
1108    3.77  NaN   NaN   NaN  NaN  NaN  
1109    5.92  NaN   NaN   NaN  NaN  NaN  
1110    7.59  NaN   NaN   NaN  NaN  NaN  
1111    3.38  NaN   NaN   NaN  NaN  NaN  
1112    5.19  NaN   NaN   NaN  NaN  NaN  
1113    6.55  NaN   NaN   NaN  NaN  NaN  
3352    4.11  NaN   NaN   NaN  NaN  NaN  
3353    5.82  NaN   NaN   NaN  NaN  NaN  

[14 rows x 23 columns]

In [353]:
main_df_away = main_df_away.sort_values(by = ['Team', 'Date'])

# Missing values
So 14 of the 4327 observations are missing team hitting data. All of them are right around the All-Star break. The days without games might have caused some kind of a glitch when we scraped the data. Remember the team hitting and bullpen data encompasses the last 30 days. We're leaning toward just dropping rows with missing data as long as we don't lose too much data.

In [354]:
main_df_home = pd.merge(main_df_home, hitting_df, how = 'left', on = ['Date', 'Team'])

In [355]:
main_df_home.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4326
Data columns (total 23 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   Team         4327 non-null   object        
 3   H_Score      4327 non-null   float64       
 4   H_Win        4327 non-null   float64       
 5   H_Team_Wins  4327 non-null   float64       
 6   H_Wins_L10   4327 non-null   float64       
 7   H_Wins_L30   4327 non-null   float64       
 8   H_Py_Wins    4327 non-null   float64       
 9   H_Run_Diff   4327 non-null   float64       
 10  H_Avg_Outs   4327 non-null   float64       
 11  H_BABIP_SP   4327 non-null   float64       
 12  H_FIP        4327 non-null   float64       
 13  H_xFIP       4327 non-null   float64       
 14  H_WAR        4327 non-null   float64       
 15  H_WHIP       4327 non-null   float64       
 16  H_SIER

# Bullpen data
Now let's read in the team bullpen data.

In [356]:
bullpen_df = pd.read_csv('BP_Data_2021-22.csv')

In [357]:
bullpen_df.head()

Unnamed: 0 Team  BABIP   FIP  xFIP  WAR  WHIP Contact% SwStr%  SIERA  RS/9  \
0          26  LAA  0.200  1.50  2.68  0.2  0.67    35.0%  35.1%   2.43  6.00   
1          27  CHW  0.400  3.92  4.58  0.0  2.25    66.7%  17.0%   4.93  0.00   
2          28  CLE  0.000  4.17  5.05  0.0  1.00    88.9%   3.3%   6.37  9.00   
3          29  MIN  0.313  2.10  3.99  0.2  1.07    68.8%  15.9%   3.55  3.86   
4          30  NYY  0.300  1.03  2.92  0.3  0.64    68.8%  16.1%   2.26  0.00   

         Date  
0  2021-04-02  
1  2021-04-02  
2  2021-04-02  
3  2021-04-02  
4  2021-04-02

In [358]:
bullpen_df.tail()

Unnamed: 0 Team  BABIP   FIP  xFIP  WAR  WHIP Contact% SwStr%  SIERA  \
10798       10824  PHI  0.356  3.48  3.65  1.2  1.51    73.6%  13.4%   3.49   
10799       10825  ARI  0.284  5.58  5.30 -0.9  1.55    78.0%  10.5%   4.78   
10800       10826  BAL  0.344  3.84  4.02  0.5  1.51    79.5%  10.2%   3.86   
10801       10827  PIT  0.317  4.50  4.38 -0.3  1.52    75.3%  11.5%   4.09   
10802       10828  COL  0.312  3.72  3.53  1.2  1.34    76.4%  10.8%   3.30   

       RS/9        Date  
10798  3.17  2022-10-06  
10799  3.63  2022-10-06  
10800  4.67  2022-10-06  
10801  4.93  2022-10-06  
10802  4.62  2022-10-06

In [359]:
bullpen_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10803 entries, 0 to 10802
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  10803 non-null  int64  
 1   Team        10803 non-null  object 
 2   BABIP       10803 non-null  float64
 3   FIP         10803 non-null  float64
 4   xFIP        10803 non-null  float64
 5   WAR         10803 non-null  float64
 6   WHIP        10803 non-null  float64
 7   Contact%    10803 non-null  object 
 8   SwStr%      10803 non-null  object 
 9   SIERA       10803 non-null  float64
 10  RS/9        10803 non-null  float64
 11  Date        10803 non-null  object 
dtypes: float64(7), int64(1), object(4)
memory usage: 1012.9+ KB


# Dropping and renaming
Dropping a few columns so we don't get out of control with dimensionality, and then we need to add a '_BP' suffix to the bullpen columns since they share a name with the some of the other columns that we've brought in. 

In [360]:
bullpen_df = bullpen_df.drop(columns = ['Unnamed: 0', 'Contact%', 'SwStr%', 'RS/9'])

In [361]:
bullpen_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10803 entries, 0 to 10802
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Team    10803 non-null  object 
 1   BABIP   10803 non-null  float64
 2   FIP     10803 non-null  float64
 3   xFIP    10803 non-null  float64
 4   WAR     10803 non-null  float64
 5   WHIP    10803 non-null  float64
 6   SIERA   10803 non-null  float64
 7   Date    10803 non-null  object 
dtypes: float64(6), object(2)
memory usage: 675.3+ KB


In [362]:
for col in list(bullpen_df.columns[1:7]):
    bullpen_df = bullpen_df.rename(columns = {col: col + '_BP'})

In [363]:
bullpen_df['Date'] = pd.to_datetime(bullpen_df['Date'])

In [364]:
bullpen_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10803 entries, 0 to 10802
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Team      10803 non-null  object        
 1   BABIP_BP  10803 non-null  float64       
 2   FIP_BP    10803 non-null  float64       
 3   xFIP_BP   10803 non-null  float64       
 4   WAR_BP    10803 non-null  float64       
 5   WHIP_BP   10803 non-null  float64       
 6   SIERA_BP  10803 non-null  float64       
 7   Date      10803 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(6), object(1)
memory usage: 675.3+ KB


In [365]:
main_df_away = pd.merge(main_df_away, bullpen_df, on = ['Date', 'Team'], how = 'left')

In [366]:
main_df_away.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4326
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   Team         4327 non-null   object        
 3   A_Score      4327 non-null   int64         
 4   A_Win        4327 non-null   int64         
 5   A_Team_Wins  4327 non-null   float64       
 6   A_Wins_L10   4327 non-null   float64       
 7   A_Wins_L30   4327 non-null   float64       
 8   A_Py_Wins    4327 non-null   float64       
 9   A_Run_Diff   4327 non-null   float64       
 10  A_Avg_Outs   4327 non-null   float64       
 11  A_BABIP_SP   4327 non-null   float64       
 12  A_FIP        4327 non-null   object        
 13  A_xFIP       4327 non-null   object        
 14  A_WAR        4327 non-null   float64       
 15  A_WHIP       4327 non-null   float64       
 16  A_SIER

# More renaming
Now we just need to add an 'A_' prefix to the hitting and bullpen columns we added to indicate that it pertains to the away team.

In [367]:
cols_to_rename = list(main_df_away.columns[18:])

In [368]:
for col in cols_to_rename:
    main_df_away = main_df_away.rename(columns = {col: 'A_' + col})

In [369]:
main_df_away.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4326
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   Team         4327 non-null   object        
 3   A_Score      4327 non-null   int64         
 4   A_Win        4327 non-null   int64         
 5   A_Team_Wins  4327 non-null   float64       
 6   A_Wins_L10   4327 non-null   float64       
 7   A_Wins_L30   4327 non-null   float64       
 8   A_Py_Wins    4327 non-null   float64       
 9   A_Run_Diff   4327 non-null   float64       
 10  A_Avg_Outs   4327 non-null   float64       
 11  A_BABIP_SP   4327 non-null   float64       
 12  A_FIP        4327 non-null   object        
 13  A_xFIP       4327 non-null   object        
 14  A_WAR        4327 non-null   float64       
 15  A_WHIP       4327 non-null   float64       
 16  A_SIER

In [370]:
main_df_home = pd.merge(main_df_home, bullpen_df, on = ['Date', 'Team'], how = 'left')

In [371]:
main_df_home.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4326
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   Team         4327 non-null   object        
 3   H_Score      4327 non-null   float64       
 4   H_Win        4327 non-null   float64       
 5   H_Team_Wins  4327 non-null   float64       
 6   H_Wins_L10   4327 non-null   float64       
 7   H_Wins_L30   4327 non-null   float64       
 8   H_Py_Wins    4327 non-null   float64       
 9   H_Run_Diff   4327 non-null   float64       
 10  H_Avg_Outs   4327 non-null   float64       
 11  H_BABIP_SP   4327 non-null   float64       
 12  H_FIP        4327 non-null   float64       
 13  H_xFIP       4327 non-null   float64       
 14  H_WAR        4327 non-null   float64       
 15  H_WHIP       4327 non-null   float64       
 16  H_SIER

In [372]:
cols_to_rename = list(main_df_home.columns[18:])

In [373]:
for col in cols_to_rename:
    main_df_home = main_df_home.rename(columns = {col: 'H_' + col})

In [374]:
main_df_home.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4326
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   Team         4327 non-null   object        
 3   H_Score      4327 non-null   float64       
 4   H_Win        4327 non-null   float64       
 5   H_Team_Wins  4327 non-null   float64       
 6   H_Wins_L10   4327 non-null   float64       
 7   H_Wins_L30   4327 non-null   float64       
 8   H_Py_Wins    4327 non-null   float64       
 9   H_Run_Diff   4327 non-null   float64       
 10  H_Avg_Outs   4327 non-null   float64       
 11  H_BABIP_SP   4327 non-null   float64       
 12  H_FIP        4327 non-null   float64       
 13  H_xFIP       4327 non-null   float64       
 14  H_WAR        4327 non-null   float64       
 15  H_WHIP       4327 non-null   float64       
 16  H_SIER

# Ballpark factors
Now we'll bring in the ballpark data that we scraped from <a href='https://baseballsavant.mlb.com/leaderboard/statcast-park-factors?type=year&year=2021&batSide=&stat=index_wOBA&condition=All&rolling=no'>MLB.com</a>. We didn't want to add too many dimensions so we just used the "Park Factor" variable that indicates whether a ballpark is more favorable to hitters or pitchers.<br>

This number is standardized so that 100 is neutral with a higher number indicating a hitters' park and a lower number indicating a pitchers' park.<br>

Now we only need to merge this data to main_df_home since we only need to match on the home team name, and the park factors are common to both teams playing in the game.



In [375]:
parks_df = pd.read_csv('Park_Factors_2021-22.csv')

In [376]:
parks_df = parks_df.drop(columns = ['Unnamed: 0'])

In [377]:
parks_df.head()

Team  Year  Park
0  COL  2021   115
1  WSH  2021   105
2  BOS  2021   104
3  BAL  2021   103
4  CIN  2021   103

In [378]:
parks_df.tail()

Team  Year  Park
55   SD  2022    95
56  OAK  2022    94
57  STL  2022    94
58  SEA  2022    94
59  TEX  2022    98

In [379]:
parks_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Team    60 non-null     object
 1   Year    60 non-null     int64 
 2   Park    60 non-null     int64 
dtypes: int64(2), object(1)
memory usage: 1.5+ KB


In [380]:
parks_df['Year'].value_counts()

2022    30
2021    30
Name: Year, dtype: int64

# Matching team codes
We thought we had already solved this problem, but perhaps that was with another dataset. We've discovered downstream that there are some team codes in the parks_df that don't match our three-letter team codes in the main dataframe. We'll use a little set theory to fix this.

In [381]:
parks_teams = list(parks_df['Team'].unique())
main_teams = list(main_df_home['Team'].unique())

In [382]:
unmatched = list(set(parks_teams) - set(main_teams))

In [383]:
unmatched

['TB', 'SD', 'SF', 'CWS', 'WSH', 'KC']

In [384]:
unmatched_reverse = list(set(main_teams) - set(parks_teams))

In [385]:
unmatched_reverse

['TBR', 'SDP', 'KCR', 'CHW', 'SFG', 'WSN']

In [386]:
teams_dict = {'TB': 'TBR', 'SD': 'SDP', 'SF': 'SFG', 'CWS': 'CHW', 'WSH': 'WSN', 'KC': 'KCR'}

In [387]:
chatGPT("""
In pandas, can I use a dictionary to change values in a column? For example, if the current value matches one of the dictionary
keys, then it's changed to the value.
""")


Yes, you can use a dictionary to map values in a column of a pandas dataframe. You can use the `map` or `replace` functions to map values in a column using a dictionary. The syntax would be something like this: 
```
df['column_name'] = df['column_name'].map(dictionary_name) 
```



In [388]:
parks_df['Team'].replace(teams_dict, inplace=True)

# Creating a Year column
Before we merge the ballpark data with the home dataframe, we need to create a column in the latter that just contains the year, since we're going to match both the team and the year, as the park factors could differ slightly between 2021 and 2022.

In [389]:
chatGPT("How can I extract a year from a datetime column in a pandas dataframe and make that another column?")



Using pandas datetime functions and DataFrame’s .apply() function, you can extract the year from a datetime column and make it its own column. For example:

# Create example DataFrame
import pandas as pd
import datetime

df = pd.DataFrame({'col_name': [datetime.datetime(2020, 3, 2, 15, 2, 8),
                                datetime.datetime(2018, 5, 1, 10, 6, 11)]})

# Add new column for year
df['year'] = df['col_name'].apply(lambda x: x.year)

# Print DataFrame
print(df)

#   col_name  year
#0 2020-03-02  2020
#1 2018-05-01  2018


In [390]:
main_df_home['Year'] = main_df_home['Date'].dt.year

In [391]:
main_df_home.head()

Date  game_id Team  H_Score  H_Win  H_Team_Wins  H_Wins_L10  \
0 2021-04-06       59  LAA      2.0    0.0          4.0         4.0   
1 2021-04-06       67  OAK      1.0    0.0          0.0         0.0   
2 2021-04-07       73  BOS      9.0    1.0          2.0         2.0   
3 2021-04-07       74  CHC      2.0    0.0          3.0         3.0   
4 2021-04-07       75  CIN     11.0    1.0          4.0         4.0   

   H_Wins_L30  H_Py_Wins  H_Run_Diff  ...  H_wOBA  H_WAR  H_ISO  H_BABIP_BP  \
0         4.0        3.0         3.0  ...   0.336    0.8  0.176       0.283   
1         0.0        0.0       -33.0  ...   0.243   -0.5  0.096       0.312   
2         2.0        2.0        -3.0  ...   0.280   -0.1  0.132       0.368   
3         3.0        3.0         1.0  ...   0.260   -0.2  0.184       0.278   
4         4.0        4.0        24.0  ...   0.409    1.9  0.246       0.163   

   H_FIP_BP  H_xFIP_BP  H_WAR_BP  H_WHIP_BP  H_SIERA_BP  Year  
0      3.62       3.65       0.2       1.27        3.66  2021  
1      6.69       5.86      -0.4       1.91        5.58  2021  
2      2.35       3.73       0.6       1.44        3.73  2021  
3      2.86       2.92       0.3       1.19        3.39  2021  
4      4.91       4.19       0.1       0.92        3.81  2021  

[5 rows x 30 columns]

In [392]:
main_df_home.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4326
Data columns (total 30 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   Team         4327 non-null   object        
 3   H_Score      4327 non-null   float64       
 4   H_Win        4327 non-null   float64       
 5   H_Team_Wins  4327 non-null   float64       
 6   H_Wins_L10   4327 non-null   float64       
 7   H_Wins_L30   4327 non-null   float64       
 8   H_Py_Wins    4327 non-null   float64       
 9   H_Run_Diff   4327 non-null   float64       
 10  H_Avg_Outs   4327 non-null   float64       
 11  H_BABIP_SP   4327 non-null   float64       
 12  H_FIP        4327 non-null   float64       
 13  H_xFIP       4327 non-null   float64       
 14  H_WAR        4327 non-null   float64       
 15  H_WHIP       4327 non-null   float64       
 16  H_SIER

In [393]:
main_df_home['Year'].value_counts()

2022    2229
2021    2098
Name: Year, dtype: int64

In [394]:
main_df_home = pd.merge(main_df_home, parks_df, on = ['Team', 'Year'], how = 'left')

In [395]:
main_df_home.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4326
Data columns (total 31 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   Team         4327 non-null   object        
 3   H_Score      4327 non-null   float64       
 4   H_Win        4327 non-null   float64       
 5   H_Team_Wins  4327 non-null   float64       
 6   H_Wins_L10   4327 non-null   float64       
 7   H_Wins_L30   4327 non-null   float64       
 8   H_Py_Wins    4327 non-null   float64       
 9   H_Run_Diff   4327 non-null   float64       
 10  H_Avg_Outs   4327 non-null   float64       
 11  H_BABIP_SP   4327 non-null   float64       
 12  H_FIP        4327 non-null   float64       
 13  H_xFIP       4327 non-null   float64       
 14  H_WAR        4327 non-null   float64       
 15  H_WHIP       4327 non-null   float64       
 16  H_SIER

# Getting back to one game, one row
We can drop the Year column we created. We also don't need the team columns, but let's hang on to them just so we can spot-check a few rows for accuracy before burning the dataframe into a csv.

In [396]:
main_df_home = main_df_home.drop(columns = ['Year'])

In [397]:
main_df_home = main_df_home.rename(columns = {'Team': 'H_Team'})

In [398]:
main_df_away = main_df_away.rename(columns = {'Team': 'A_Team'})

In [399]:
main_df = pd.merge(main_df_away, main_df_home, on = ['Date', 'game_id'])

In [400]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4327 entries, 0 to 4326
Data columns (total 57 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4327 non-null   datetime64[ns]
 1   game_id      4327 non-null   int64         
 2   A_Team       4327 non-null   object        
 3   A_Score      4327 non-null   int64         
 4   A_Win        4327 non-null   int64         
 5   A_Team_Wins  4327 non-null   float64       
 6   A_Wins_L10   4327 non-null   float64       
 7   A_Wins_L30   4327 non-null   float64       
 8   A_Py_Wins    4327 non-null   float64       
 9   A_Run_Diff   4327 non-null   float64       
 10  A_Avg_Outs   4327 non-null   float64       
 11  A_BABIP_SP   4327 non-null   float64       
 12  A_FIP        4327 non-null   object        
 13  A_xFIP       4327 non-null   object        
 14  A_WAR        4327 non-null   float64       
 15  A_WHIP       4327 non-null   float64       
 16  A_SIER

In [401]:
rows_with_missing = main_df.isnull().any(axis = 1)

In [402]:
missing_df = main_df[rows_with_missing]

In [403]:
len(missing_df)

14

# Only 14 missing rows
We could go back and get the data for the missing rows, but that would be a headache. We still have data from 88.7 percent of the games played in 2021 and 2022.

In [404]:
main_df = main_df.dropna()

In [405]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4313 entries, 0 to 4326
Data columns (total 57 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4313 non-null   datetime64[ns]
 1   game_id      4313 non-null   int64         
 2   A_Team       4313 non-null   object        
 3   A_Score      4313 non-null   int64         
 4   A_Win        4313 non-null   int64         
 5   A_Team_Wins  4313 non-null   float64       
 6   A_Wins_L10   4313 non-null   float64       
 7   A_Wins_L30   4313 non-null   float64       
 8   A_Py_Wins    4313 non-null   float64       
 9   A_Run_Diff   4313 non-null   float64       
 10  A_Avg_Outs   4313 non-null   float64       
 11  A_BABIP_SP   4313 non-null   float64       
 12  A_FIP        4313 non-null   object        
 13  A_xFIP       4313 non-null   object        
 14  A_WAR        4313 non-null   float64       
 15  A_WHIP       4313 non-null   float64       
 16  A_SIER

In [406]:
main_df = main_df.reset_index(drop = True)

In [407]:
main_df = main_df.sort_values(by = 'Date')

In [409]:
main_df.iloc[824, :]

Date           2021-06-18 00:00:00
game_id                       1032
A_Team                         CIN
A_Score                          2
A_Win                            0
A_Team_Wins                     35
A_Wins_L10                       7
A_Wins_L30                      17
A_Py_Wins                       33
A_Run_Diff                      -2
A_Avg_Outs                      14
A_BABIP_SP                   0.357
A_FIP                         4.88
A_xFIP                        7.53
A_WAR                            0
A_WHIP                        1.93
A_SIERA                       5.86
A_RS/9                        5.79
A_OPS                        0.744
A_BABIP                      0.283
A_wOBA                       0.323
A_WAR                          2.5
A_ISO                        0.176
A_BABIP_BP                   0.331
A_FIP_BP                      5.07
A_xFIP_BP                     4.28
A_WAR_BP                       0.1
A_WHIP_BP                     1.51
A_SIERA_BP          

# Target variable
Now let's derive our target variable. Then we'll run a couple of tests to verify.

In [412]:
main_df['Winner'] = np.where(main_df['H_Score'] > main_df['A_Score'], 'H', 'A')

In [415]:
main_df['Winner'].value_counts()

H    2324
A    1989
Name: Winner, dtype: int64

In [413]:
main_df[(main_df['Winner'] == 'H') & (main_df['H_Win'] == 0)]

Empty DataFrame
Columns: [Date, game_id, A_Team, A_Score, A_Win, A_Team_Wins, A_Wins_L10, A_Wins_L30, A_Py_Wins, A_Run_Diff, A_Avg_Outs, A_BABIP_SP, A_FIP, A_xFIP, A_WAR, A_WHIP, A_SIERA, A_RS/9, A_OPS, A_BABIP, A_wOBA, A_WAR, A_ISO, A_BABIP_BP, A_FIP_BP, A_xFIP_BP, A_WAR_BP, A_WHIP_BP, A_SIERA_BP, H_Team, H_Score, H_Win, H_Team_Wins, H_Wins_L10, H_Wins_L30, H_Py_Wins, H_Run_Diff, H_Avg_Outs, H_BABIP_SP, H_FIP, H_xFIP, H_WAR, H_WHIP, H_SIERA, H_RS/9, H_OPS, H_BABIP, H_wOBA, H_WAR, H_ISO, H_BABIP_BP, H_FIP_BP, H_xFIP_BP, H_WAR_BP, H_WHIP_BP, H_SIERA_BP, Park, Winner]
Index: []

[0 rows x 58 columns]

In [416]:
len(main_df[(main_df['Winner'] == 'H') & (main_df['H_Win'] == 1)])

2324

In [417]:
main_df[(main_df['Winner'] == 'A') & (main_df['A_Win'] == 0)]

Empty DataFrame
Columns: [Date, game_id, A_Team, A_Score, A_Win, A_Team_Wins, A_Wins_L10, A_Wins_L30, A_Py_Wins, A_Run_Diff, A_Avg_Outs, A_BABIP_SP, A_FIP, A_xFIP, A_WAR, A_WHIP, A_SIERA, A_RS/9, A_OPS, A_BABIP, A_wOBA, A_WAR, A_ISO, A_BABIP_BP, A_FIP_BP, A_xFIP_BP, A_WAR_BP, A_WHIP_BP, A_SIERA_BP, H_Team, H_Score, H_Win, H_Team_Wins, H_Wins_L10, H_Wins_L30, H_Py_Wins, H_Run_Diff, H_Avg_Outs, H_BABIP_SP, H_FIP, H_xFIP, H_WAR, H_WHIP, H_SIERA, H_RS/9, H_OPS, H_BABIP, H_wOBA, H_WAR, H_ISO, H_BABIP_BP, H_FIP_BP, H_xFIP_BP, H_WAR_BP, H_WHIP_BP, H_SIERA_BP, Park, Winner]
Index: []

[0 rows x 58 columns]

In [418]:
len(main_df[(main_df['Winner'] == 'A') & (main_df['A_Win'] == 1)])

1989

In [419]:
main_df = main_df.drop(columns = ['game_id', 'A_Team', 'A_Score', 'A_Win', 'H_Team', 'H_Score', 'H_Win'])

In [420]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4313 entries, 1877 to 4312
Data columns (total 51 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         4313 non-null   datetime64[ns]
 1   A_Team_Wins  4313 non-null   float64       
 2   A_Wins_L10   4313 non-null   float64       
 3   A_Wins_L30   4313 non-null   float64       
 4   A_Py_Wins    4313 non-null   float64       
 5   A_Run_Diff   4313 non-null   float64       
 6   A_Avg_Outs   4313 non-null   float64       
 7   A_BABIP_SP   4313 non-null   float64       
 8   A_FIP        4313 non-null   object        
 9   A_xFIP       4313 non-null   object        
 10  A_WAR        4313 non-null   float64       
 11  A_WHIP       4313 non-null   float64       
 12  A_SIERA      4313 non-null   object        
 13  A_RS/9       4313 non-null   float64       
 14  A_OPS        4313 non-null   float64       
 15  A_BABIP      4313 non-null   float64       
 16  A_w

In [421]:
chatGPT("How do I turn a column into the index in a pandas dataframe?")



You can set a column as the index in a pandas dataframe by calling the set_index() method on the dataframe object and specifying the column name as the argument. For example:

# set the 'Name' column as the index
df.set_index('Name')


In [423]:
main_df = main_df.set_index('Date')

In [424]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4313 entries, 2021-04-06 to 2022-10-05
Data columns (total 50 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   A_Team_Wins  4313 non-null   float64
 1   A_Wins_L10   4313 non-null   float64
 2   A_Wins_L30   4313 non-null   float64
 3   A_Py_Wins    4313 non-null   float64
 4   A_Run_Diff   4313 non-null   float64
 5   A_Avg_Outs   4313 non-null   float64
 6   A_BABIP_SP   4313 non-null   float64
 7   A_FIP        4313 non-null   object 
 8   A_xFIP       4313 non-null   object 
 9   A_WAR        4313 non-null   float64
 10  A_WHIP       4313 non-null   float64
 11  A_SIERA      4313 non-null   object 
 12  A_RS/9       4313 non-null   float64
 13  A_OPS        4313 non-null   float64
 14  A_BABIP      4313 non-null   float64
 15  A_wOBA       4313 non-null   float64
 16  A_WAR        4313 non-null   float64
 17  A_ISO        4313 non-null   float64
 18  A_BABIP_BP   4313 non-null   f

In [425]:
filepath = r'C:\Users\Owner\Sports Betting\MLB_Game_Outcome\main_dataframe.csv'
main_df.to_csv(filepath)